src/rngs/adapter/reseeding.rs

```diff
@@ -177,26 +191,51 @@ where R: BlockRngCore + SeedableRng,
         })
     }
 
+    fn is_forked(&self, global_fork_counter: usize) -> bool {
+        // In theory, on 32-bit platforms, it is possible for
+        // `global_fork_counter` to wrap around after ~4e9 forks.
+        //
+        // This check will detect a fork in the normal case where
+        // `fork_counter < global_fork_counter`, and also when the difference
+        // between both is greater than `isize::MAX` (wrapped around).
+        //
+        // It will still fail to detect a fork if there have been more than
+        // `isize::MAX` forks, without any reseed in between. Seems unlikely
+        // enough.
+        (self.fork_counter.wrapping_sub(global_fork_counter) as isize) < 0
+    }
+
     #[inline(never)]
     fn reseed_and_generate(&mut self,
-                           results: &mut <Self as BlockRngCore>::Results)
+                           results: &mut <Self as BlockRngCore>::Results,
+                           global_fork_counter: usize)
     {
-        trace!("Reseeding RNG after {} generated bytes",
-               self.threshold - self.bytes_until_reseed);
-        let threshold = if let Err(e) = self.reseed()  {
+        if self.is_forked(global_fork_counter) {
+            info!("Fork detected, reseeding RNG");
+        } else {
+            // FIXME: the amount logged here may not be correct if the intial
+            // `bytes_until_reseed` was set by a delay instead of `threshold`.
+            trace!("Reseeding RNG after {} generated bytes",
```

Comment:
We don't need to include the byte count here at all. In fact since it can be incorrect, it might be better not to.